### ISSUE TITLE
Rancher CLI inconsistency with agent ip address field names vs whats returned in API

### ISSUE THREAD

**Issue by aemneina** [rancher/rancher#9838]
**Rancher versions:**
rancher/server: v1.6.7
rancher/agent:

rancher --env dev hosts ls --format {{.Host.Id}}#{{.Host.Hostname}}#{{.Host.State}}#{{.Host.InstanceIds}}#{{.Host.agentIpAddress}}#{{.Host.Labels}} 
FATA[0000] template: :1:77: executing "" at <.Host.agentIpAddress>: can't evaluate field agentIpAddress in type client.Host

It cannot find the agentIpAddress field while it is copy/pasted from the API view for the host, but the following works:

rancher --env dev hosts ls --format {{.Host.Id}}#{{.Host.Hostname}}#{{.Host.State}}#{{.Host.InstanceIds}}#{{.Host.AgentIpAddress}}#{{.Host.Labels}} 
1h265#USE1-RANCHER-DEV-A01#active#[1i269632 1i269633 1i269639 1i269634 1i269640 1i283426 1i283469 1i283470 1i283471 1i283472 1i283522 1i283524 1i283525 ….

The difference is just in Capital Letter the field name starts from .Host.agentIpAddress vs .Host.AgentIpAddress 
Same is true for “state” field - didn’t work until specified with Capital “S”

I think somewhere string comparison is not done right and something like equalsIgnoreCase(string, string) is missing.

**Comment by aiwantaozi** [comment]
@yasker @aemneina is this by design? All fields use CamelCase to format the result.

**Comment by vincent99** [comment]
The fields in templates are from the Golang objects, as they are Golang templates, and all public properties start with a capital letter.  So they are all `TitleCased`.

**Comment by yasker** [comment]
@aemneina So where do you get `Host.agentIpAddress`?

**Comment by aiwantaozi** [comment]
thanks, understand we use "text/template" to format Golang objects, it a little wired when use Json format, some key start with capital letter, and some start with lower-case letter. See the picture below. Is this also work as we expected?

![image](https://user-images.githubusercontent.com/5936271/30581547-267f379e-9d53-11e7-8d8f-85547b0129a1.png)

**Comment by yasker** [comment]
Confirmed with @vincent99 , the `ID` and `Host` are wrappers around the API, so it's not a part of API response. I think we should be OK here.

But still, the format of `Host.xxx` template should be in CamelCase, so we may need to update the doc for that.

**Comment by aiwantaozi** [comment]
When we use the custom output format, some fields in result starts with lower-case, but we need to use the capital letter in the query condition. Because we use go template "https://golang.org/pkg/text/template/" to parse. For the golang object, the template could not parse any none exported fields, so the output condition only can use the capital letter. If we want to unify the query condition and the result, we may need to change not to use this template and implement in another way.

**Comment by yasker** [comment]
@aiwantaozi 

Docker is using the [CamelCase](https://docs.docker.com/engine/admin/formatting/#template-functions).

But kubectl [is not](https://kubernetes.io/docs/tasks/access-application-cluster/list-all-running-container-images/).

Kubectl declares it's using Go templates too. So we need to figure out why it doesn't need to use CamelCase.

**Comment by yasker** [comment]
Logan's PR converts all template to Title format, so it will match Golang text template. So, user can use the value with a lowercase first letter in the template.

LGTM. What do you think? @vincent99 @aemneina @will-chan

**Comment by vincent99** [comment]
Doesn't that break backwards compatibility with existing uses?

**Comment by yasker** [comment]
It only accept TitleCase before. So previous error out cases will be fine
now. And previous working ones are TitleCase already, so should be no
change.

**Comment by loganhz** [comment]
@vincent99 
I have tested `rancher --env dev hosts ls --format {{.Host.Id}}#{{.Host.Hostname}}#{{.Host.State}}#{{.Host.InstanceIds}}#{{.Host.agentIpAddress}}#{{.Host.Labels}}`
and `rancher --env dev hosts ls --format {{.Host.Id}}#{{.Host.Hostname}}#{{.Host.State}}#{{.Host.InstanceIds}}#{{.Host.AgentIpAddress}}#{{.Host.Labels}}`

it looks good.
Please let me know if there is any other usage I should try.

**Comment by loganhz** [comment]
`rancher --env dev hosts ls --format "hello {{.Host.Id}}"`
returns 
`Hello 1h1`
Is that good if we just title string inside `{{}}`

**Comment by aemneina** [comment]
can we introduce a cli flag for behavior change? that way current use still works, new method requires flag?

**Comment by StrongMonkey** [comment]
@aemneina `--ignoreCase`? 😄 why can't we just use `AgentIpAddress`?

**Comment by loganhz** [comment]
@aemneina yes, we can. It will be great if you can provide the flag name and usage description for it. :)  thanks

**Comment by yasker** [comment]
@aemneina TitleCase is fine with Vince, let me double check and merge it.

**Comment by vincent99** [comment]
I don't get why a flag is needed here.

[The PR](https://github.com/rancher/cli/pull/96/files) only seems to address the case of trying to select a field that starts with an uppercase character using a format starting with a lowercase (format: `agentIpAddress`, field: `AgentIpAddress`).  What about the other way around (format: `AgentIpAddress`, field: `agentIpAddress`), or arbitrary other places in the string (format `AgentIPAddress` vs `AgentIpAddress`)?

Seems like it should be either:

a) the user has to put in exactly the right key in the format.  If there are keys which are currently inconsistent with the expected naming we fix them and populate both the original and corrected key in the context for backwards compatibility

or

b) Make everything case-insensitive in both directions, i.e. normalize all the format and context keys to all lowercase before interpreting.  There are no objects with keys differing only in case (because that would be terrible).  So this should not make any key ambiguous.

**Comment by aemneina** [comment]
Option b sounds good to me.

**Comment by loganhz** [comment]
As we need to normalize the context keys to lowercase, the keys in output also are changed to lowercase. Does it matter?

For example,

before:
input: `rancher hosts ls --format '{{.Host.Info.osInfo}}'`  
output: `map[dockerVersion:Docker version 1.12.6, build 78d1802 kernelVersion:4.4.0 operatingSystem:Ubuntu 14.04.5 LTS]`

after:
input: `rancher hosts ls --format '{{.Host.Info.osInfo}}'`  
output: `map[dockerversion:Docker version 1.12.6, build 78d1802 kernelversion:4.4.0 operatingsystem:Ubuntu 14.04.5 LTS]`

**Comment by vincent99** [comment]
Selecting an object/array and letting go stringify it seems a bit odd.  But the result is ugly..

To make that work you could instead leave the data alone, walk over all the keys and construct a map of lowercased -> original paths:

`.host.info.osinfo.dockerversion` -> `.Host.Info.osInfo.DockerVersion`

Then take each format key, lowercase it, and replace with the matching map value.  (If no match is found then that key doesn't exist).

**Comment by loganhz** [comment]
thanks @vincent99 
but I guess it won't work for the case `{{(index .Host.PublicEndpoints 2).hostId}}` which should be `{{(index .Host.PublicEndpoints 2).HostId}}`

**Comment by loganhz** [comment]
Hi @aemneina 

It looks like we still cannot make it case insensitive with go-template. It will introduce some side effect as we discussed here.

Can we add a flag (--json) which makes user be able to use the field while it is copy/pasted from the API view? So that it won’t impact current user who is using the title case already.

For example,

rancher --env dev hosts ls --format {{.Host.Id}}#{{.Host.Hostname}}#{{.Host.State}}#{{.Host.InstanceIds}}#{{.Host.**agentIpAddress**}}#{{.Host.Labels}} **--json**

**Comment by yasker** [comment]
What do you think @vincent99 ? It's difficult to support two kinds of format automatically due to the flexible of go-template.
